solarforecastarbiter/metrics/evaluator.py

```diff
@@ -0,0 +1,175 @@
+    obs_method = obs_context['fill_method']
+    if obs_method in context.supported_fill_functions():
+        fill_func = context._FILL_FUNCTIONS_MAP[obs_method]
+        obs_values = fill_func(observations.value, observations.quality_flag)
+    else:
+        obs_values = observations.values
+
+    # Preprocessing forecasts
+    fx_context = metrics_context['preprocessing']['forecasts']
+    fx_method = fx_context['fill_method']
+    if fx_method in context.supported_fill_functions():
+        fill_func = context._FILL_FUNCTIONS_MAP[fx_method]
+        fx_values = fill_func(forecasts)
+    else:
+        fx_values = forecasts
+
+    # Copy preprocessed timeseries to result
+    if metrics_context['results']['timeseries']['observations']:
+        result['timeseries']['observations'] = obs_values
+
+    if metrics_context['results']['timeseries']['forecasts']:
+        result['timeseries']['forecasts'] = fx_values
+
+    # Calculate metrics
+    for metric, do_calc in metrics_context['metrics'].items():
+        if do_calc and metric in context.supported_metrics():
+
+            # For all
+            metric_func = context._METRICS_MAP[metric]
+            result['metrics']['total'][metric] = metric_func(obs_values,
+                                                             fx_values)
+
+            # Group by calculations
+            group_context = metrics_context['results']['groupings']
+            for groupby, do_group in group_context.items():
+                if do_group and groupby in context.supported_groupings():
+
+                    if groupby not in result['metrics'].keys():
+                        result['metrics'][groupby] = {}
+
+                    result['metrics'][groupby][metric] = \
+                        evaluate_by_group(obs_values,
+                                          fx_values,
+                                          groupby,
+                                          metric_func)
+
+    return result
+
+
+def evaluate_by_group(observation, forecast, groupby, metric_func):
+    """Evaluate the performance according to the groupby.
+
+    Parameters
+    ----------
+    observation : pd.Series
+    forecast : pd.Series
+    groupby : string
+        one of the supported types of groupings
+    metric_func : function
+        a py:func:`solarforecastarbiter.metrics` metrics function
+
+    Returns
+    -------
+    pd.Series :
+
+    Raises
+    ------
+    SfaMetricsConfigError if `groupby` is not a supported groupby type.
+    """
+    result = None
+
+    # Verify timestamps
+    assert np.array_equal(observation.index, forecast.index)
+
+    # Group into single DataFrame
+    df = pd.DataFrame({'observation': observation,
+                       'forecast': forecast},
+                      index=observation.index)
+
+    # Determine proper groupby
+    if groupby == 'month':
+        df_group = df.groupby(df.index.month)
+    elif groupby == 'dow':
```

Comment:
to me, it's more clear if we stick with the pandas names: weekday, hour.

Reply:
Makes sense. Done